taskchampion/taskchampion/src/server/cloud/server.rs

```diff
@@ -0,0 +1,1130 @@
+        expected.mock_set_latest(v3);
+        assert_eq!(server.unencrypted(), expected.unencrypted());
+    }
+
+    #[test]
+    fn cleanup_old_versions_no_snapshot() {
+        let mut server = make_server();
+        let (v1, v2, v3) = (Uuid::new_v4(), Uuid::new_v4(), Uuid::new_v4());
+        server.mock_add_version(v1, v2, 200, b"second");
+        server.mock_add_version(v2, v3, 300, b"third");
+        server.mock_set_latest(v3);
+
+        server.cleanup().unwrap();
+
+        // Nothing is deleted.
+        let mut expected = server.empty_copy();
+        expected.mock_add_version(v1, v2, 200, b"second");
+        expected.mock_add_version(v2, v3, 300, b"third");
+        expected.mock_set_latest(v3);
+        assert_eq!(server.unencrypted(), expected.unencrypted());
+    }
+
+    #[test]
+    fn cleanup_old_versions_with_snapshot() {
+        let mut server = make_server();
+        let (v1, v2, v3) = (Uuid::new_v4(), Uuid::new_v4(), Uuid::new_v4());
+        let (v4, v5, v6) = (Uuid::new_v4(), Uuid::new_v4(), Uuid::new_v4());
+        server.mock_add_version(v1, v2, 200, b"second");
+        server.mock_add_version(v2, v3, 300, b"third");
+        server.mock_add_version(v3, v4, 1400, b"fourth");
+        server.mock_add_version(v4, v5, 1500, b"fifth");
+        server.mock_add_snapshot(v5, 1501, b"snap 1");
+        server.mock_add_version(v5, v6, 1600, b"sixth");
+        server.mock_set_latest(v6);
+
+        server.cleanup().unwrap();
+
+        let mut expected = server.empty_copy();
+        expected.mock_add_version(v3, v4, 1400, b"fourth"); // Not old enough to be deleted.
+        expected.mock_add_version(v4, v5, 1500, b"fifth");
+        expected.mock_add_snapshot(v5, 1501, b"snap 1");
+        expected.mock_add_version(v5, v6, 1600, b"sixth");
+        expected.mock_set_latest(v6);
+        assert_eq!(server.unencrypted(), expected.unencrypted());
+    }
+
+    #[test]
+    fn cleanup_old_versions_newer_than_snapshot() {
+        let mut server = make_server();
+        let (v1, v2, v3) = (Uuid::new_v4(), Uuid::new_v4(), Uuid::new_v4());
+        let (v4, v5, v6) = (Uuid::new_v4(), Uuid::new_v4(), Uuid::new_v4());
+        server.mock_add_version(v1, v2, 200, b"second");
+        server.mock_add_version(v2, v3, 300, b"third");
+        server.mock_add_snapshot(v3, 301, b"snap 1");
+        server.mock_add_version(v3, v4, 400, b"fourth");
+        server.mock_add_version(v4, v5, 500, b"fifth");
+        server.mock_add_version(v5, v6, 600, b"sixth");
+        server.mock_set_latest(v6);
+
+        server.cleanup().unwrap();
+
+        let mut expected = server.empty_copy();
+        expected.mock_add_snapshot(v3, 301, b"snap 1");
+        expected.mock_add_version(v3, v4, 400, b"fourth");
+        expected.mock_add_version(v4, v5, 500, b"fifth");
+        expected.mock_add_version(v5, v6, 600, b"sixth");
+        expected.mock_set_latest(v6);
+        assert_eq!(server.unencrypted(), expected.unencrypted());
+    }
```

Comment:
If you built up `expected` before running `cleanup` you could clearly demonstrate that `cleanup()` is what is actually doing the work. For example:

```rs
        assert_ne!(server.unencrypted(), expected.unencrypted());
        server.cleanup().unwrap();
        assert_eq!(server.unencrypted(), expected.unencrypted());
```

Reply:
Good idea - thanks!